File: JASSjr_index.pl
#!/usr/bin/env perl

# Copyright (c) 2024 Vaughan Kitchen
# Minimalistic BM25 search engine.

use v5.38;
use strict;
use warnings;
use builtin qw(true false);
no warnings 'experimental';

# Make sure we have one parameter, the filename
die "Usage: $0 <infile.xml>" if scalar @ARGV != 1;

my %vocab; # the in-memory index
my @doc_ids; # the primary keys
my @length_vector; # hold the length of each document

my $docid = -1;
my $document_length = 0;
my $push_next = false; # is the next token the primary key?

while (<>) {
	# A token is either an XML tag '<'..'>' or a sequence of alpha-numerics.
	# TREC <DOCNO> primary keys have a hyphen in them
	foreach (m/[a-zA-Z0-9][a-zA-Z0-9-]*|<[^>]*>/g) {
		# If we see a <DOC> tag then we're at the start of the next document
		if ($_ eq '<DOC>') {
			# Save the previous document length
			push @length_vector, $document_length if $docid != -1;
			# Move on to the next document
			$docid += 1;
			$document_length = 0;
			say "$docid documents indexed" if $docid % 1000 == 0;
		}
		# if the last token we saw was a <DOCNO> then the next token is the primary key
		if ($push_next) {
			push @doc_ids, $_;
			$push_next = false;
		}
		$push_next = true if $_ eq '<DOCNO>';
		# Don't index XML tags
		next if substr($_, 0, 1) eq '<';

		# lower case the string
		my $token = lc $_;
		# truncate any long tokens at 255 charactes (so that the length can be stored first and in a single byte)
		$token = substr $token, 0, 255;

		# add the posting to the in-memory index
		$vocab{$token} = [] if not exists $vocab{$token};
		my $postings_list = $vocab{$token};
		if (scalar @{$postings_list} == 0 || @{$postings_list}[-2] ne $docid) {
			push @{$postings_list}, $docid, 1;
		} else {
			@{$postings_list}[-1] += 1;
		}

		# Compute the document length
		$document_length += 1;
	}
}

# If we didn't index any documents then we're done.
exit if $docid == -1;

# Save the final document length
push @length_vector, $document_length;

# tell the user we've got to the end of parsing
say "Indexed @{[$docid + 1]} documents. Serialising...";

# Save the final document length
open my $docids_fh, '>', 'docids.bin' or die;
foreach (@doc_ids) {
	print $docids_fh "$_\n";
}

open my $postings_fh, '>:raw', 'postings.bin' or die;
open my $vocab_fh, '>:raw', 'vocab.bin' or die;

while (my ($term, $postings) = each %vocab) {
	my @postings = @{$postings};

	# write the postings list to one file
	my $where = tell $postings_fh;
	print $postings_fh pack 'l*', @postings;

	# write the vocabulary to a second file (one byte length, string, '\0', 4 byte where, 4 byte size)
	print $vocab_fh pack 'Ca*xll', length($term), $term, $where, scalar @postings * 4;
}

# store the document lengths
open my $lengths_fh, '>:raw', 'lengths.bin' or die;
print $lengths_fh pack 'l*', @length_vector;

# clean up
close $docids_fh;
close $postings_fh;
close $vocab_fh;
close $lengths_fh;

File: JASSjr_index.lua
#!/usr/bin/env luajit

-- Copyright (c) 2024 Vaughan Kitchen
-- Minimalistic BM25 search engine.

-- Allow strings to be accessed like an array
getmetatable('').__index = function(str, i) return string.sub(str, i, i) end
-- Allow strings to be "called" to access substr
getmetatable('').__call = string.sub

-- Lua5.3 includes pack but we are targeting Lua5.1
function pack_int(num)
	return string.char(
		bit.band(num, 0xFF),
		bit.rshift(bit.band(num, 0xFF00), 8),
		bit.rshift(bit.band(num, 0xFF0000), 16),
		bit.rshift(bit.band(num, 0xFF000000), 24)
	)
end

-- Make sure we have one parameter, the filename
if #arg ~= 1 then
	print(string.format("Usage: %s <infile.xml>", arg[0]))
	os.exit()
end

local vocab = {} -- the in-memory index
local doc_ids = {} -- the primary keys
local doc_lengths = {} -- hold the length of each document

local docid = -1
local document_length = 0
local push_next = false -- is the next token the primary key?

local fh = assert(io.open(arg[1]))
for line in fh:lines() do
	-- A token is either an XML tag '<'..'>' or a sequence of alpha-numerics
	-- TREC <DOCNO> primary keys have a hyphen in them
	-- Lua patterns aren't full regex. We can't do alternation on groups
	-- As our document is fairly well formed we should be fine
	for token in string.gmatch(line, "<?/?[a-zA-Z0-9][a-zA-Z0-9-]*") do
		-- If we see a <DOC> tag then we're at the start of the next document
		if token == "<DOC" then
			-- Save the previous document length
			if docid ~= -1 then
				table.insert(doc_lengths, document_length)
			end
			-- Move on to the next document
			docid = docid + 1
			document_length = 0
			if docid % 1000 == 0 then
				print(string.format("%d documents indexed", docid))
			end
		end
		-- If the last token we saw was a <DOCNO> then the next token is the primary key
		if push_next then
			table.insert(doc_ids, token)
			push_next = false
		end
		if token == "<DOCNO" then
			push_next = true
		end
		-- Don't index XML tags
		if token[1] ~= "<" then
			-- TODO handle / at start of string
			-- Lowercase the string
			token = string.lower(token)

			-- Truncate any long tokens at 255 characters (so that the length can be stored first and in a single byte)
			token = token(1, 255)

			-- Add the posting to the in-memory index
			local postings = vocab[token] or {}
			local postings_len = #postings -- TODO make this more efficient? (is O(log(n))
			if postings_len == 0 or postings[postings_len - 1] ~= docid then
				-- If the docno for this occurence has changed then create a new <d,tf> pair
				table.insert(postings, docid)
				table.insert(postings, 1)
			else
				postings[postings_len] = postings[postings_len] + 1
			end
			vocab[token] = postings

			-- Compute the document length
			document_length = document_length + 1
		end
	end
end

-- If we didn't index any documents then we're done
if docid == -1 then
	os.exit()
end

-- Save the final document length
table.insert(doc_lengths, document_length)

-- Tell the user we've got to the end of parsing
print(string.format("Indexed %d documents. Serialising...", docid + 1))

-- Store the primary keys
local docids_fh = io.open("docids.bin", "w")
for _, docid in ipairs(doc_ids) do
	docids_fh:write(docid, "\n")
end

local postings_fh = io.open("postings.bin", "w")
local vocab_fh = io.open("vocab.bin", "w")
for term, postings in pairs(vocab) do
	-- Write the postings list to one file
	local where = postings_fh:seek()
	for _, post in ipairs(postings) do
		postings_fh:write(pack_int(post))
	end

	-- Write the vocabulary to a second file (one byte length, string, '\0', 4 byte where, 4 byte size)
	vocab_fh:write(string.char(string.len(term)), term, "\0", pack_int(where), pack_int(#postings * 4))
end

-- Store the document lengths
local lengths_fh = io.open("lengths.bin", "w")
for _, length in ipairs(doc_lengths) do
	lengths_fh:write(pack_int(length))
end

-- Clean up
docids_fh:close()
postings_fh:close()
vocab_fh:close()
lengths_fh:close()
